wsmux/stream.go

```diff
@@ -0,0 +1,237 @@
+package wsmux
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+const (
+	DefaultCapacity = 1024
+)
+
+type stream struct {
+	id        uint32
+	m         sync.Mutex
+	c         *sync.Cond
+	b         *buffer
+	unblocked uint32
+
+	endErr error
+
+	closed       bool
+	remoteClosed bool
+	accepted     chan struct{}
+
+	session *Session
+
+	//deadlines
+	writeDeadline time.Time
+	readDeadline  time.Time
+}
+
+func newStream(id uint32, session *Session) *stream {
+	str := &stream{
+		id:        id,
+		b:         newBuffer(DefaultCapacity),
+		unblocked: 0,
+
+		endErr: nil,
+
+		closed:       false,
+		remoteClosed: false,
+		accepted:     make(chan struct{}),
+
+		session: session,
+	}
+
+	str.c = sync.NewCond(&str.m)
+	return str
+}
+
+func (s *stream) unblock(read uint32) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	defer s.session.logger.Printf("unblock broadcasted : stream %d", s.id)
+	s.unblocked += read
+}
+
+func (s *stream) push(buf []byte) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	defer s.session.logger.Printf("push broadcasted : stream %d", s.id)
+	_, err := s.b.Write(buf)
+	s.endErr = err
+}
+
+func (s *stream) accept(read uint32) {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	s.unblocked += read
+
+	select {
+	case <-s.accepted:
+	default:
+		close(s.accepted)
+	}
+}
+
+func (s *stream) getBufLen() int {
+	s.m.Lock()
+	defer s.m.Unlock()
+	return s.b.Len()
+}
+
+func (s *stream) setRemoteClosed() {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+	s.remoteClosed = true
+}
+
+func (s *stream) isClosed() bool {
+	s.m.Lock()
+	defer s.m.Unlock()
+	return s.closed
+}
+
+func (s *stream) isRemoteClosed() bool {
+	s.m.Lock()
+	defer s.m.Unlock()
+	return s.remoteClosed
+}
+
+func (s *stream) LocalAddr() net.Addr {
+	return s.session.conn.LocalAddr()
+}
+
+func (s *stream) RemoteAddr() net.Addr {
+	return s.session.conn.RemoteAddr()
+}
+
+func (s *stream) SetReadDeadline(t time.Time) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.readDeadline = t
+	return nil
+}
+
+func (s *stream) SetWriteDeadline(t time.Time) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	s.writeDeadline = t
+	return nil
+}
+
+func (s *stream) SetDeadline(t time.Time) error {
+	if err := s.SetReadDeadline(t); err != nil {
+		return err
+	}
+	if err := s.SetWriteDeadline(t); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Close closes the stream
+func (s *stream) Close() error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	defer s.c.Broadcast()
+
+	s.closed = true
+	if err := s.session.send(newFinFrame(s.id)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Read reads bytes from the stream
+func (s *stream) Read(buf []byte) (int, error) {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	s.session.logger.Printf("stread %d: read requested", s.id)
+
+	timeout, timer := getTimeoutAndTimer(s.readDeadline)
```

Comment:
Store the timer on `stream` probably instead of storing `readDeadline`...
If the timer triggers then set `stream.readDeadlineExceeded = true` and broadcast`

Whenever, `Read()` is ready to return it then clears to the timer... and `SetReadDeadline` sets the timer again (and sets `stream.readDeadlineExceeded = false`).
This way you don't constantly recreate the timer... you stop/start it and move it's deadline, but that's all you do...

Also you currently have a deadlock... if nobody does a `c.Broadcast()` and the timer times out, then we never get to the `case <-timeout:` case, because the function is deadlocked at `c.Wait()`.